mirgecom/fluid.py

```diff
@@ -0,0 +1,62 @@
+""":mod:`mirgecom.fluid` provides common utilities for fluid simulation.
+
+.. autofunction:: compute_local_velocity_gradient
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+import numpy as np  # noqa
+from meshmode.dof_array import DOFArray  # noqa
+from mirgecom.euler import ConservedVars
+
+
+def compute_local_velocity_gradient(discr, cv: ConservedVars):
+    r"""
+    Compute the cell-local gradient of fluid velocity.
+
+    Computes the cell-local gradient of fluid velocity from:
+
+    .. math::
+
+        \nabla{v_i} = \frac{1}{\rho}(\nabla(\rho{v_i})-v_i\nabla{\rho}),
+
+    where $v_i$ is ith velocity component.
+
+    Parameters
+    ----------
+    discr: grudge.eager.EagerDGDiscretization
+        the discretization to use
+    cv: mirgecom.euler.ConservedVars
+        the fluid conserved variables
+    Returns
+    -------
+    numpy.ndarray
+        object array of :class:`~meshmode.dof_array.DOFArray`
+        representing $\partial_j{v_i}$.
+    """
+    dim = discr.dim
+    velocity = cv.momentum/cv.mass
+    dmass = discr.grad(cv.mass)
+    dmom = [discr.grad(cv.momentum[i]) for i in range(dim)]
+    return [(dmom[i] - velocity[i]*dmass)/cv.mass for i in range(dim)]
```

Comment:
Doc says it returns a numpy array, but it's currently returning a list.

Could maybe do something along the lines of:

```suggestion
    dmom = np.array([discr.grad(cv.momentum[i]) for i in range(dim)], dtype=object)
    return dmom - np.outer(velocity, dmass)/cv.mass
```
?

Reply:
I had to tweak this slightly back to a clunkier form (ae88854), but it may have just been because the whole array needed to be divided by `cv.mass`. At least the returned structure is right now.  I'll try to bring it back closer to your expression after addressing the other comments.

Reply:
(d7f3dc9)